realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,279 @@
+     *
+     * @throws IllegalArgumentException If the new configuration isn't valid.
+     */
+    private void validateConfiguration(RealmConfiguration newConfiguration) {
+        // != takes no time, check it first
+        if (newConfiguration == configuration || newConfiguration.hashCode() == configuration.hashCode()) {
+            // Same configurations
+            return;
+        }
+
+        // Check that encryption keys aren't different
+        if (!Arrays.equals(configuration.getEncryptionKey(), newConfiguration.getEncryptionKey())) {
+            throw new IllegalArgumentException(DIFFERENT_KEY_MESSAGE);
+        }
+
+        // Check schema versions are the same
+        if (configuration.getSchemaVersion() != newConfiguration.getSchemaVersion()) {
+            throw new IllegalArgumentException(String.format("Configurations cannot have different schema versions " +
+                            "if used to open the same file. %d vs. %d", configuration.getSchemaVersion(),
+                    newConfiguration.getSchemaVersion()));
+        }
+
+        // Check that schema is the same
+        RealmProxyMediator cachedSchema = configuration.getSchemaMediator();
+        RealmProxyMediator schema = newConfiguration.getSchemaMediator();
+        if (!cachedSchema.equals(schema)) {
+            throw new IllegalArgumentException("Two configurations with different schemas are trying to open " +
+                    "the same Realm file. Their schema must be the same: " + newConfiguration.getPath());
+        }
+
+        // Check if the durability is the same
+        SharedGroup.Durability cachedDurability = configuration.getDurability();
+        SharedGroup.Durability newDurability = newConfiguration.getDurability();
+        if (!cachedDurability.equals(newDurability)) {
+            throw new IllegalArgumentException("A Realm cannot be both in-memory and persisted. Two conflicting " +
+                    "configurations pointing to " + newConfiguration.getPath() + " are being used.");
+        }
+
+        // Should never get here
+        throw new IllegalArgumentException(
+                "The new configuration pointing to " + newConfiguration.getPath() +
+                        " has a different hash code with a cached configuration. " +
+                        "The new configuration's hashcode: " + newConfiguration.hashCode() +
+                        "The cached configuration's hashcode: " + configuration.hashCode() + ".");
+    }
+
+    /**
+     * Runs the callback function with the total reference count of {@link Realm} and {@link DynamicRealm} who refer to
+     * the given {@link RealmConfiguration}.
+     *
+     * @param configuration the {@link RealmConfiguration} of {@link Realm} and {@link DynamicRealm}.
+     * @param callback the callback will be executed with the global reference count.
+     */
+    static synchronized void invokeWithGlobalRefCount(RealmConfiguration configuration, Callback callback) {
```

Comment:
If it's used only for test purpose why not use Reflection? + you're adding a new public interface Callback?